src/node_contractor.rs

```diff
@@ -56,15 +56,13 @@ pub fn handle_shortcuts<F>(
 ) where
     F: FnMut(&mut PreparationGraph, Shortcut),
 {
-    dijkstra.avoid_node(node);
     for i in 0..graph.in_edges[node].len() {
         for j in 0..graph.out_edges[node].len() {
             let weight = graph.in_edges[node][i].weight + graph.out_edges[node][j].weight;
-            dijkstra.set_max_weight(weight);
             let in_node = graph.in_edges[node][i].adj_node;
             let out_node = graph.out_edges[node][j].adj_node;
             let best_weight = dijkstra.calc_weight(graph, in_node, out_node);
-            if best_weight.is_none() {
+            if best_weight.is_some() && best_weight.unwrap() >= weight {
```

Comment:
I do not understand this yet: you deleted the line that ignores the contracting node for the witness search. This means that in_node and out_node are always connected by the direct path in_node->node->out_node and the weight of this path is `weight` here. Then we do the witness search (local dijkstra) and because here you do not explicitly avoid `node`  this dijkstra always finds the direct path *or* it finds a shorter path. This means that whatever happens it holds that `best_weight <= weight`. Under no circumstances it can be that `best_weight > weight`. So your condition here really means that we are always introducing a shortcut whenever the weight of the witness path equals the weight of the direct path. This is probably an error and it should be just the other way around (`best_weight.unwrap() <= weight`)?. But in this case I do not see how this method is supposed to *reduce* the amount of shortcuts compared to the current version in master.

Reply:
your approach is to exclude the node from the graph and check how "essential" this node is for the local nodes. If it is essential not path is found. Then you create a shortcut.
my approach is turning this around. You leave all nodes like the graph represents them. Then you calculate a path from neighbors before to the neighbors after it. So you definitely know that the worst path you can get is the path directly using the contracting node. If there is a path that is somehow around your contracting node, then you do not have to create a shortcut, since there is a better way which should be used for querys. This path will be handled in the future.
Having here an `==` or `>=` does exactly the same.

Reply:
Yes, exactly. And yes I got it wrong. It has to be: `if best_weight>=weight introduce the shortcut` or to make it more clear `if best_weight==weight introduce the shortcut`, meaning: if the direct path is really our best option, introduce the shortcut.

Reply:
yes exactly.

Reply:
So no the problem with the approach you are suggesting seems to be the case we have in `calc_shortcuts_witness` where the alternative path has the exact same weight as the direct path. In this case we are introducing a shortcut, even though the alternative path is just as good.

Reply:
i updated the code to check for `exactly`

Reply:
> So now the problem with the approach you are suggesting seems to be the case we have in `calc_shortcuts_witness` where the alternative path has the exact same weight as the direct path. In this case we are introducing a shortcut, even though the alternative path is just as good.

yes true, and here the node ordering will be important.
in my implementation i am also using the heuristic `amount of contracted neighbors` if we apply this, then it is likely, that in the future the the "node 3" will be contracted before "0" or "2"
if the contraction would contract "node 3", this will create another shortcut from 0 -> 2 and then your de-duplication would fix it.

Reply:
i can think of examples, where this will create more amount of shortcuts. Specially if the ordering of the nodes is given.

Reply:
Ok sure, but not introducing shortcuts that are not needed in the node contraction routine is probably the safest bet here.

Reply:
oh yes I agree there

Reply:
i will try implementing your idea with sorting

Reply:
considering the example above:
how do we determine, that there is a path from 0 to 3? which is still cheaper?
specially when we have this example:
```
0 -> 1 -> 2
|   /
4  /
| /
3
```

Reply:
should we still resolve the path and look if any other neighboring node is in the resolved path?

Reply:
oh there is no path by using `avoid_node`